source/frontend/src/features/admin/financial-codes/add/AddFinancialCodeForm.test.tsx

```diff
@@ -0,0 +1,155 @@
+  onSave: jest.fn(),
+  onCancel: jest.fn(),
+  onError: jest.fn(),
+  onSuccess: jest.fn(),
+};
+
+describe('AddFinancialCode form', () => {
+  const setup = (renderOptions: RenderOptions = {}) => {
+    const utils = render(
+      <AddFinancialCodeForm
+        validationSchema={mockProps.validationSchema}
+        onSave={mockProps.onSave}
+        onSuccess={mockProps.onSuccess}
+        onCancel={mockProps.onCancel}
+        onError={mockProps.onError}
+      />,
+      {
+        ...renderOptions,
+      },
+    );
+    return { ...utils };
+  };
+
+  beforeAll(() => {
+    // Lock the current datetime as our snapshot has date-dependent fields
+    jest.useFakeTimers('modern');
+    jest.setSystemTime(new Date('04 Dec 2015 10:15:00 GMT').getTime());
+  });
+
+  beforeEach(() => {
+    // reset mock yup validation between tests
+    mockProps.validationSchema = Yup.object().shape({});
+    jest.resetAllMocks();
+  });
+
+  afterAll(() => {
+    // back to reality...
+    jest.useRealTimers();
+  });
+
+  it('renders as expected', () => {
+    const { asFragment } = setup();
+    expect(asFragment()).toMatchSnapshot();
+  });
+
+  it('calls onCancel when form is not dirty and cancel is clicked', async () => {
+    setup();
+
+    const cancelButton = screen.getByText('Cancel');
+    await act(() => userEvent.click(cancelButton));
+
+    expect(mockProps.onCancel).toHaveBeenCalled();
+  });
+
+  it('displays modal when form is dirty and cancel is clicked', async () => {
+    setup();
+
+    const description = document.querySelector(`input[name="description"]`) as HTMLInputElement;
+    const cancelButton = screen.getByText('Cancel');
+    await act(() => userEvent.paste(description, `another description`));
+    await act(() => userEvent.click(cancelButton));
+
+    expect(screen.getByText('Unsaved Changes')).toBeVisible();
+    expect(await screen.findByDisplayValue('another description')).toBeVisible();
+  });
+
+  it('calls onCancel when form is dirty, cancel is clicked and modal is confirmed', async () => {
+    setup();
+
+    const description = document.querySelector(`input[name="description"]`) as HTMLInputElement;
+    const cancelButton = screen.getByText('Cancel');
+    await act(() => userEvent.paste(description, `another description`));
+    await act(() => userEvent.click(cancelButton));
+    expect(screen.getByText('Unsaved Changes')).toBeVisible();
+    const confirmButton = screen.getByText('Confirm');
+    await act(() => userEvent.click(confirmButton));
+
+    expect(mockProps.onCancel).toHaveBeenCalled();
+    expect(description).toHaveTextContent('');
+  });
+
+  it('does not call onCancel when form is dirty, cancel is clicked but modal is not confirmed', async () => {
+    setup();
+
+    const textbox = document.querySelector(`input[name="description"]`) as HTMLInputElement;
+    const cancelButton = screen.getByText('Cancel');
+    await act(() => userEvent.paste(textbox, `another description`));
```

Comment:
haven't seen this before, whats the advantage of using paste here? Should we update our fill input method to use this?

Reply:
Personal preference here. I kinda dislike the fill method a bit as I have found issues in the past or if you forget to specify "textarea" as the last parameter it doesn't work for text areas. I would prefer if we could try to use userEvent instead. 
And yes, we should update/refactor our userEvent

Reply:
Also I use "paste" instead of `userEvent.type` because I'm not interested in waiting for 10 (or 20) keystrokes for eg: "sample description".